src/graphics/Screen.cpp

```diff
@@ -402,43 +409,170 @@ static bool shouldDrawMessage(const meshtastic_MeshPacket *packet)
-    static char tempBuf[237];
+    std::string result;
 
-    const meshtastic_MeshPacket &mp = devicestate.rx_text_message;
-    meshtastic_NodeInfoLite *node = nodeDB->getMeshNode(getFrom(&mp));
-    // LOG_DEBUG("drawing text message from 0x%x: %s\n", mp.from,
-    // mp.decoded.variant.data.decoded.bytes);
+    // Using a copy to preserve the original text
+    char *textCopy = new char[text.length() + 1];
+    std::strcpy(textCopy, text.c_str());
 
-    // Demo for drawStringMaxWidth:
-    // with the third parameter you can define the width after which words will
-    // be wrapped. Currently only spaces and "-" are allowed for wrapping
-    display->setTextAlignment(TEXT_ALIGN_LEFT);
-    display->setFont(FONT_SMALL);
-    if (config.display.displaymode == meshtastic_Config_DisplayConfig_DisplayMode_INVERTED) {
-        display->fillRect(0 + x, 0 + y, x + display->getWidth(), y + FONT_HEIGHT_SMALL);
-        display->setColor(BLACK);
+    char *line = std::strtok(textCopy, "\n");
+    while (line != nullptr) {
+        uint16_t preferredBreakpoint = 0;
+        uint16_t strWidth = 0;
+        std::string currentLine;
+
+        size_t lineLength = std::strlen(line);
+        for (uint16_t i = 0; i < lineLength; ++i) {
+            char character[2] = {line[i], '\0'};
+            strWidth += display->getStringWidth(character, 1, true);
+
+            if (line[i] == ' ' || line[i] == '-' || line[i] == '/') {
+                preferredBreakpoint = i + 1;
+            }
+
+            if (strWidth >= maxLineWidth && preferredBreakpoint > 0) {
+                result += currentLine.substr(0, preferredBreakpoint) + '\n';
+                line += preferredBreakpoint;
+                i = -1;                         // Reset the index for the next line
+                lineLength = std::strlen(line); // Update length for the remaining string
+                currentLine.clear();
+                strWidth = 0;            // Reset width for the new line
+                preferredBreakpoint = 0; // Reset breakpoint
+                continue;
+            }
+
+            currentLine += line[i];
+        }
+
+        result += currentLine;                        // Add the last line or the whole line if never exceeded maxLineWidth
+        if (*(line + currentLine.length()) != '\0') { // If not at the end of the text, add a newline
+            result += '\n';
+        }
+
+        line = std::strtok(nullptr, "\n"); // Get next line if any
     }
 
-    uint32_t seconds = sinceReceived(&mp);
+    delete[] textCopy; // Clean up the copied text
+
+    return result;
+}
+
+// Format message for displaying as text
+std::string createMessageString(OLEDDisplay *display, const meshtastic_Message msg)
+{
+    uint32_t seconds = sinceReceived(&msg);
     uint32_t minutes = seconds / 60;
     uint32_t hours = minutes / 60;
     uint32_t days = hours / 24;
+    std::string timestamp = screen->drawTimeDelta(days, hours, minutes, seconds);
```

Comment:
Rather than "time since received", could this display the local-time that the message was received (where possible)? Otherwise this information quickly becomes invalid on E-Ink displays, which won't refresh this info until a new message arrives.

Reply:
Displaying the time and date would take up a lot of screen space which is why I went with this approach. For E-Ink displays, you can manually refresh by doing any action. For now, that'll have to be the workaround.

Reply:
If I'm being honest, I'm particularly satisfied with this idea as a workaround. I think there's definitely a workable solution here somewhere. Seeing how this behavior is really only important to E-Ink devices, it's absolutely possible to implement two different pieces of code here, wrapped with the `USE_EINK` macro.

If it helps, I can have a play around and see if I can come up with code for this?

Having spend a tiny bit more time testing the PR now, I'm not actually sure this threaded-messaging idea is super helpful for devices with smaller screens. Realistically, there's only enough space to display one message at a time anyway. It might actually be an idea to disable this threaded-messaging feature entirely for some variants?

Reply:
I should've mentioned that I didn't mean for the E-Ink workaround to be a permanent solution. I'm just trying to get to a minimum viable product for such a large update since I think, even in a not perfect state, it's a big step forwards for users. I don't have an E-Ink device to test on so determining if anything I code even works will not be possible. If you'd like to play around with it, that would be great.

Threaded messaging is still useful on small displays since you can scroll to see older messages. I do think this feature should be disabled for some variants as I mentioned [here](https://github.com/meshtastic/firmware/pull/3784#discussion_r1594862401).

Reply:
Resolving since now E-Ink devices will use the old message frame.